### ISSUE TITLE
Error: Debug Failure. createKeyForReference received unexpected expression type

### ISSUE THREAD

**Issue by rslinckx** [microsoft/pyright#3695]
**Describe the bug**
Latest release (1.1.260) trigger an internal error

**To Reproduce**
```
$ npx pyright@1.1.260
[snip]
An internal error occurred while binding file "/Users/r.slinckx/.pyenv/versions/3.10.1/lib/python3.10/multiprocessing/util.py": Error: Debug Failure. createKeyForReference received unexpected expression type
    at createKeyForReference (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/codeFlowTypes.ts:240:9)
    at createKeyForReference (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/codeFlowTypes.ts:218:25)
    at C.visitAssignment (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/binder.ts:849:50)
    at C.visitNode (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:132:29)
    at C.walk (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:98:37)
    at forEach (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:107:22)
    at Array.forEach (<anonymous>)
    at C.walkMultiple (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:105:15)
    at C.walk (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:100:18)
    at C._walkStatementsAndReportUnreachable (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/binder.ts:2583:22)
    at C.visitSuite (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/binder.ts:345:14)
    at C.visitNode (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:330:29)
    at C.walk (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:98:37)
    at callback (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/binder.ts:1281:18)
    at C._trackCodeFlowExpressions (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/binder.ts:3236:9)
    at C.visitIf (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/binder.ts:1265:48)
    at C.visitNode (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:184:29)
    at C.walk (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:98:37)
    at C._walkStatementsAndReportUnreachable (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/binder.ts:2583:22)
    at C.visitSuite (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/binder.ts:345:14)
    at C.visitNode (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:330:29)
    at C.walk (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:98:37)
    at Object.callback (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/binder.ts:545:22)
    at C._bindDeferred (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/binder.ts:4140:22)
    at C.bindModule (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/binder.ts:298:14)
    at callback (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/sourceFile.ts:1197:28)
    at s.timeOperation (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/common/timing.ts:44:28)
    at callback (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/sourceFile.ts:1184:38)
    at t.LogTracker.log (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/common/logTracker.ts:31:20)
    at t.SourceFile.bind (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/sourceFile.ts:1181:33)
    at O._bindFile (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/program.ts:931:34)
    at callback (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/program.ts:976:22)
    at subtractFromTime (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/common/timing.ts:56:13)
    at importLookup (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/program.ts:975:41)
    at Object.resolveAliasDeclaration (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/aliasDeclarationUtils.ts:62:28)
    at resolveAliasDeclaration (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:19204:33)
    at getInferredTypeOfDeclaration (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:18945:30)
    at forEach (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:19362:36)
    at Array.forEach (<anonymous>)
    at getEffectiveTypeOfSymbolForUsage (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:19309:15)
    at getTypeOfName (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:3848:43)
    at Object.getTypeOfExpression (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:933:30)
    at isCallNoReturn (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/codeFlowEngine.ts:1275:40)
    at isFlowNodeReachableRecursive (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/codeFlowEngine.ts:992:44)
    at Object.isFlowNodeReachable (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/codeFlowEngine.ts:1068:20)
    at Object.isAfterNodeReachable (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:2564:29)
    at isAfterNodeReachable (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/codeFlowEngine.ts:1466:26)
    at isFunctionNoReturn (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/codeFlowEngine.ts:1452:51)
    at callback (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/codeFlowEngine.ts:1349:25)
    at doForEachSubtype (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeUtils.ts:473:9)
    at isCallNoReturn (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/codeFlowEngine.ts:1277:13)
    at getTypeFromFlowNode (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/codeFlowEngine.ts:366:48)
    at Object.getTypeFromCodeFlow (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/codeFlowEngine.ts:912:20)
    at getFlowTypeOfReference (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:17822:25)
    at getTypeOfName (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:3883:48)
    at getTypeOfExpression (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:933:30)
    at getDeclaredTypeForExpression (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:2114:21)
    at evaluateTypesForAssignmentStatement (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:14255:38)
    at evaluateTypesForStatement (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:17621:25)
    at callback (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:19075:17)
    at evaluateTypeForSubnode (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:17743:13)
    at getInferredTypeOfDeclaration (/Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:19074:32)
    at /Users/r.slinckx/.npm/_npx/95971/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:19362:36
    at Array.forEach (<anonymous>)
```

I also get a similar error on a file inside my project, probably the same thing ?

Not sure how i can help further ?

**Comment by last-partizan** [comment]
@rslinckx you should always provide minimal reproduction.

here it is for this case:

```python
from typing import Any, Dict


placeholders = {}
fields: Dict[str, Any] = {}

for k, v in placeholders.items():
    fields[k].widget.attrs["placeholder"] = v
```

**Comment by erictraut** [comment]
Oh, sorry about that. Yes, this is a similar manifestation of the regression I fixed in 1.1.260. I've introduced a fix that covers this case as well. It will be included in the next release.

In the meantime, you can work around the problem by breaking up the expression.
```python
temp = fields[k].widget
temp.attrs["placeholder"] = v
```

**Comment by erictraut** [comment]
This is included in pyright 1.1.261, which I just published. It will also be included in an upcoming release of pylance.
